src/coreclr/jit/rangecheckcloning.cpp

```diff
@@ -0,0 +1,520 @@
+        : GenTreeVisitor(compiler)
+        , m_stmt(stmt)
+        , m_boundsChks(bndChkLocations)
+    {
+    }
+
+    fgWalkResult PostOrderVisit(GenTree** use, GenTree* user)
+    {
+        GenTree* node = *use;
+        if (node->OperIs(GT_BOUNDS_CHECK))
+        {
+            m_boundsChks->Push(BoundCheckLocation(m_stmt, node->AsBoundsChk(), user));
+        }
+        return fgWalkResult::WALK_CONTINUE;
+    }
+};
+
+// -----------------------------------------------------------------------------
+// optRangeCheckCloning: The main entry point for the range check cloning phase.
+//    This phase scans all the blocks in the method and groups the bounds checks
+//    in each block by the "Base Index and Length" pairs (VNs). Then it picks up
+//    the largest group and clones the block to have fast and slow paths in order
+//    to optimize the bounds checks in the fast path.
+//    See the overview at the top of the file and the comments in the optRangeCheckCloning_DoClone
+//    function for more details.
+//
+// Return Value:
+//    The status of the phase after the transformation.
+//
+PhaseStatus Compiler::optRangeCheckCloning()
+{
+    if (!doesMethodHaveBoundsChecks())
+    {
+        JITDUMP("Current method has no bounds checks\n");
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+    if (lvaHaveManyLocals(0.75))
+    {
+        // If we're close to running out of locals, we should be conservative
+        JITDUMP("Too many locals - bail out.\n");
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+    const bool preferSize = opts.jitFlags->IsSet(JitFlags::JIT_FLAG_SIZE_OPT);
+    if (preferSize)
+    {
+        // The optimization comes with a codegen size increase
+        JITDUMP("Optimized for size - bail out.\n");
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+    bool modified = false;
+
+    // An array to keep all the bounds checks in the block
+    // Strictly speaking, we don't need this array and can group the bounds checks
+    // right as we walk them, but this helps to improve the TP/Memory usage
+    // as many blocks don't have enough bounds checks to clone anyway.
+    ArrayStack<BoundCheckLocation> bndChkLocations(getAllocator(CMK_RangeCheckCloning));
+
+    // A map to group the bounds checks by the base index and length VNs
+    BoundsCheckInfoMap bndChkMap(getAllocator(CMK_RangeCheckCloning));
+
+    for (BasicBlock* block = fgFirstBB; block != nullptr; block = block->Next())
+    {
+        if (!block->HasFlag(BBF_MAY_HAVE_BOUNDS_CHECKS))
+        {
+            // TP optimization - skip blocks that *likely* don't have bounds checks
+            continue;
+        }
+
+        if (block->isRunRarely())
+        {
+            continue;
+        }
+
+        if (!block->KindIs(BBJ_ALWAYS, BBJ_RETURN, BBJ_COND))
+        {
+            // For now, we support only simple blocks
+            continue;
+        }
```

Comment:
What support is missing for this?

Reply:
I think nothing, your suggestion with `HasTerminator` should handle it